### ISSUE TITLE
Pylance incorrectly reports an unknown member error when using pattern matching

### ISSUE THREAD

**Issue by ferrata** [microsoft/pylance-release#2198]
## Environment data

-   Language Server version: 2021.12.2
-   OS and version: Windows 10 (version 2004, OS Build 19041.1415)
-   Python version (& distribution if applicable, e.g. Anaconda): Python 3.10.0 64-bit

## Expected behaviour

Should not report an error

## Actual behaviour

Reports an error as below

<img width="630" alt="image" src="https://user-images.githubusercontent.com/31716295/147747799-0ec32d4b-5c32-4e9f-8883-1dc9f0aad15d.png">

## Logs

```
Experiment 'pythonaacf' is active
Experiment 'pythonTensorboardExperiment' is active
Experiment 'pythonSurveyNotificationcf' is active
Experiment 'PythonPyTorchProfiler' is active
Experiment 'pythonDeprecatePythonPath' is active
> conda info --json
> C:\Python39\python.exe ~\.vscode\extensions\ms-python.python-2021.12.1559732655\pythonFiles\interpreterInfo.py
> ~\AppData\Local\Programs\Python\Python310\python.exe ~\.vscode\extensions\ms-python.python-2021.12.1559732655\pythonFiles\interpreterInfo.py
Python interpreter path: ~\AppData\Local\Programs\Python\Python310\python.exe
> conda --version
Starting Pylance language server.
```

```
[Info  - 5:21:33 AM] Pylance language server 2021.12.2 (pyright 211ebe44) starting
[Info  - 5:21:33 AM] Server root directory: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist
[Info  - 5:21:33 AM] No configuration file found.
[Info  - 5:21:33 AM] No pyproject.toml file found.
[Info  - 5:21:33 AM] Setting pythonPath for service "py": "C:\Users\sasha\AppData\Local\Programs\Python\Python310\python.exe"
[Warn  - 5:21:33 AM] stubPath c:\Users\sasha\my\so\py\typings is not a valid directory.
[Info  - 5:21:33 AM] Assuming Python version 3.10
[Info  - 5:21:33 AM] Assuming Python platform Windows
Search paths for c:\Users\sasha\my\so\py
  c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib
  c:\Users\sasha\my\so\py
  c:\Users\sasha\my\so\py\typings
  c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stubs\...
  c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\bundled\stubs
  c:\Users\sasha\AppData\Local\Programs\Python\Python310\DLLs
  c:\Users\sasha\AppData\Local\Programs\Python\Python310\Lib
  c:\Users\sasha\AppData\Local\Programs\Python\Python310
  c:\Users\sasha\AppData\Local\Programs\Python\Python310\Lib\site-packages
[Info  - 5:21:33 AM] Searching for source files
[Info  - 5:21:33 AM] Found 1 source file
[Info  - 5:21:33 AM] Background analysis(1) root directory: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist
[Info  - 5:21:33 AM] Background analysis(1) started
Background analysis message: setConfigOptions
Background analysis message: setImportResolver
Background analysis message: ensurePartialStubPackages
[FG] parsing: c:\Users\sasha\my\so\py\hasattr.py (12ms)
[FG] parsing: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 2ms] (70ms)
[FG] binding: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\builtins.pyi (29ms)
[FG] binding: c:\Users\sasha\my\so\py\hasattr.py (1ms)
Background analysis message: setTrackedFiles
Background analysis message: markAllFilesDirty
Background analysis message: setFileOpened
Background analysis message: getSemanticTokens full
[BG(1)] getSemanticTokens full at c:\Users\sasha\my\so\py\hasattr.py ...
[BG(1)]   parsing: c:\Users\sasha\my\so\py\hasattr.py (12ms)
[BG(1)]   parsing: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\builtins.pyi [fs read 3ms] (72ms)
[BG(1)]   binding: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\builtins.pyi (29ms)
[BG(1)]   binding: c:\Users\sasha\my\so\py\hasattr.py (0ms)
[BG(1)]   parsing: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\typing.pyi [fs read 1ms] (19ms)
[BG(1)]   binding: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\typing.pyi (7ms)
[BG(1)]   parsing: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi [fs read 1ms] (7ms)
[BG(1)]   binding: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\_typeshed\__init__.pyi (1ms)
[BG(1)]   parsing: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\types.pyi [fs read 0ms] (7ms)
[BG(1)]   binding: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\types.pyi (2ms)
[BG(1)]   parsing: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\typing_extensions.pyi [fs read 0ms] (2ms)
[BG(1)]   binding: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\typing_extensions.pyi (0ms)
[BG(1)]   parsing: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\abc.pyi [fs read 1ms] (1ms)
[BG(1)]   binding: c:\Users\sasha\.vscode\extensions\ms-python.vscode-pylance-2021.12.2\dist\typeshed-fallback\stdlib\abc.pyi (0ms)
[BG(1)] getSemanticTokens full at c:\Users\sasha\my\so\py\hasattr.py (200ms)
Background analysis message: analyze
[BG(1)] analyzing: c:\Users\sasha\my\so\py\hasattr.py ...
[BG(1)]   checking: c:\Users\sasha\my\so\py\hasattr.py (3ms)
[BG(1)] analyzing: c:\Users\sasha\my\so\py\hasattr.py (3ms)
Background analysis message: resumeAnalysis
```

## Code Snippet / Additional information

```python

class X:
    title = "1"
class Y:
    name = "2"
class Z:
    name = "3"

for (i, r) in enumerate([X(), Y(), Z()]): # type of r: X | Y | Z

    match r:
        case object(title=_): print(r.title) # error
        case object(name=_): print(r.name) # error
        case _: print('no match')
```

**Comment by erictraut** [comment]
From a type-checking perspective, the current behavior is arguably correct. The type of `r` is `X | Y | Z`, which means that `r` contains an object that is instantiated from class `X`, `Y`, `Z` _or some subclass of one of those three classes_. 

When pyright (the type checker that underlies pylance) analyzes this code, it attempts to "narrow" the type of `r` based on the patterns provided in the case statements. The starting type of `r` is `X | Y | Z`. All three of these subtypes match an `object` pattern in the first case statement, so no narrowing is possible based on the `object` class. You might think that further narrowing is possible based on the `title` argument pattern, but the match subject could theoretically be a subclass of `Y` or `Z` that adds a `title` attribute, so these cannot be safely eliminated during narrowing.

Your usage of the match statement is pretty atypical, which is why this issue hasn't been raised previously. Normally, class patterns should use the types `X`, `Y` and `Z` rather than using `object` to match against attribute values. Is there a reason you didn't use this more typical approach?

```python
    match r:
        case X():
            print(r.title)
        case Y() | Z():
            print(r.name)
        case _:
            print('no match')
```

This approach doesn't suffer from the same issue I mentioned above, and pyright is able to safely narrow the type of `r` based on these patterns.

**Comment by erictraut** [comment]
Here's what I can do. It is safe to perform type narrowing based on class argument patterns if the class is marked `@final` because no derived classes are possible in this case.

I've updated the logic in pyright to perform type narrowing in this case. This will be included in the next release of pylance. You will still need to change your code above to add `@final` class decorators to `X`, `Y` and `Z` if you want this to work.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2022.1.0, which we've just released. You can find the changelog here:  [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202210-5-january-2022)
